docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,415 @@
+# ADR 001: ABCI++ Adoption
+
+## Changelog
+
+- 03/03/2022: Initial Commit
+
+## Context
+
+Among other things, ABCI++ enables arbitrary logic to be performed by the application to create and verify proposal blocks.
+
+We need this functionality in order for block producers to
+
+- Pick a square size / Fill the data square efficiently
+- Malleate user transactions
+- Separate rollup block data from the stateful portion of the transaction
+- Use the appropriate signature for the selected block size
+- Follow the non-interactive default rules (not done here)
+- Create celestia specific data hash by erasuring the block data
+
+We also need this functionality for validators to verify that
+
+- Messages that are paid for by users are actually included in the block
+- The block producer doesn’t also include messages that are not paid for
```

Comment:
But they would be allowed to include zero fee messages, right?

Reply:
yes, that's a good point! It would be more accurate to say that it has a corresponding `MsgPayForMessage`, which is not necessarily the same as paying for it